_posts/2020-07-09-implementing-retry-with-resilience4j.md

```diff
@@ -0,0 +1,386 @@
+    String metricName = meter.getId().getTag("kind");
+    Double metricValue = StreamSupport.stream(meter.measure().spliterator(), false).
+                                        filter(m -> m.getStatistic().name().equals("COUNT")).
+                                        findFirst().
+                                        map(m -> m.getValue()).
+                                        orElse(0.0);
+    System.out.println(desc + " - " + metricName + ": " + metricValue);
+};
+meterRegistry.forEachMeter(meterConsumer);
+```
+
+Here's some sample output:
+
+```
+The number of successful calls without a retry attempt - successful_without_retry: 4.0
+The number of failed calls without a retry attempt - failed_without_retry: 0.0
+The number of failed calls after a retry attempt - failed_with_retry: 0.0
+The number of successful calls after a retry attempt - successful_with_retry: 6.0
+```
+
+Of course, in a real application, we would export the data to a monitoring system and view it on a dashboard.
+
+## Gotchas and Good Practices When Retrying
+
+Often services provide client libraries or SDKs which have a built-in retry mechanism. This is especially true for cloud services. For example, Azure CosmosDB and Azure Service Bus provide client libraries with a built-in retry facility. They allow applications to set retry policies to control the retry behavior.
+
+In such cases, it's better to use the built-in retries rather than coding our own. If we do need to write our own, we should disable the built-in default retry policy - otherwise, it could lead to nested retries where each attempt from the application causes multiple attempts from the client library.
+
+Some cloud services document transient error codes. Azure SQL for example, provides a list of error codes for which it expects database clients to retry. It's good to check if service providers have such lists before deciding to add retry for a particular operation. 
+
+Another good practice is to maintain the values in `RetryConfig` like maximum attempts, wait time, retryable error codes, exceptions  as a configuration outside our service. If we discover new transient errors or we need to tweak the interval between attempts, we can make the change without building and redeploying the service.
+
+## Conclusion
+
+In this article, we learned what Resilience4j is and how we can use it's Retry module to make our applications resilient to temporary errors. We looked at the different ways to configure retries and some examples for deciding between the various approaches. We learned some good practices to follow when implementing retries and the importance of collecting and analyzing retry metrics.
+
+You can play around with a complete application illustrating these ideas using the code [on GitHub](https://github.com/thombergs/code-examples/tree/master/resilience4j/retry). 
+
+## References
+
+1. Reslience4j Retry module documentation: https://resilience4j.readme.io/docs/retry
```

Comment:
I moved this link to the first mention of the retry module. It's best for blog posts to link to things directly from the text because many readers won't make it to the end :)